File: ITMO.Python/PythonProject/Demo02/benchmark/csharp/csharp/Engine/BigIntMathematics.cs
using System.Numerics;

namespace csharp.Engine
{
    internal class BigIntMathematics: Mathematics<BigInteger>
    {
        public BigIntMathematics() { }

        public override BigInteger[,] MultiplyMatricesParallel(BigInteger[,] matrixA, BigInteger[,] matrixB, int threadCount)
        {
            int rowA = matrixA.GetLength(0);
            int rowB = matrixB.GetLength(0);
            int colB = matrixB.GetLength(1);

            BigInteger[,] result = new BigInteger[rowA, colB];
            Parallel.For(0, rowA - 1, new ParallelOptions { MaxDegreeOfParallelism = threadCount }, row => {
                for (int col = 0; col < colB; col++)
                {
                    for (int idx = 0; idx < rowB; idx++)
                    {
                        result[row, col] = BigInteger.Add(result[row, col],
                                                          BigInteger.Multiply(matrixA[row, idx],
                                                                              matrixB[idx, col]));
                    }
                }
            });
            return result;
        }

        public override BigInteger[,] MultiplyMatrices(BigInteger[,] matrixA, BigInteger[,] matrixB)
        {
            int rowA = matrixA.GetLength(0);
            int rowB = matrixB.GetLength(0);
            int colB = matrixB.GetLength(1);

            BigInteger[,] result = new BigInteger[rowA, colB];
            for (int row = 0; row < rowA; row++)
            {
                for (int col = 0; col < colB; col++)
                {
                    for (int idx = 0; idx < rowB; idx++)
                    {
                        result[row, col] = BigInteger.Add(result[row, col],
                                                          BigInteger.Multiply(matrixA[row, idx],
                                                                              matrixB[idx, col]));
                    }
                }
            }
            return result;
        }

        public override BigInteger[,] AddMatrices(BigInteger[,] matrixA, BigInteger[,] matrixB)
        {
            int rowA = matrixA.GetLength(0);
            int colA = matrixA.GetLength(1);

            for (int row = 0; row < rowA; row++)
            {
                for (int col = 0; col < colA; col++)
                {
                    matrixA[row, col] = BigInteger.Add(matrixA[row, col], matrixB[row, col]);
                }
            }
            return matrixA;
        }

        public override BigInteger[,] MultiplyByScalar(BigInteger[,] matrix, BigInteger scalar)
        {
            int rowCount = matrix.GetLength(0);
            int colCount = matrix.GetLength(1);

            for (int row = 0; row < rowCount; row++)
            {
                for (int col = 0; col < colCount; col++)
                {
                    matrix[row, col] = BigInteger.Multiply(matrix[row, col], scalar);
                }
            }

            return matrix;
        }
    }
}

File: ITMO.Python/PythonProject/Demo02/benchmark/java/src/Engine/BigIntMathematics.java
package Engine;

import java.math.BigInteger;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class BigIntMathematics implements IMathematics<BigInteger> {

    @Override
    public BigInteger[][] transposeMatrix(BigInteger[][] matrix) {
        int rows = matrix.length;
        int columns = matrix[0].length;

        BigInteger[][] transposedMatrix = new BigInteger[columns][rows];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < columns; col++) {
                transposedMatrix[col][row] = matrix[row][col];
            }
        }

        return transposedMatrix;
    }

    @Override
    public BigInteger[][] multiplyMatrices(BigInteger[][] matrixA, BigInteger[][] matrixB) {
        BigInteger[][] result = new BigInteger[matrixA.length][matrixB[0].length];
        for (int row = 0; row < matrixA.length; row++) {
            for (int col = 0; col < matrixB[0].length; col++) {
                for (int idx = 0; idx < matrixB.length; idx++) {
                    result[row][col] = result[row][col] == null ?
                            matrixA[row][idx].multiply(matrixB[idx][col])
                            : result[row][col].add(matrixA[row][idx].multiply(matrixB[idx][col]));
                }
            }
        }
        return result;
    }

    @Override
    public BigInteger[][] multiplyMatricesParallel(BigInteger[][] matrixA, BigInteger[][] matrixB, int threadCount) {
        BigInteger[][] result = new BigInteger[matrixA.length][matrixB[0].length];

        ExecutorService executor = Executors.newFixedThreadPool(threadCount);
        CountDownLatch latch = new CountDownLatch(matrixA.length);
        for (int idx = 0; idx < matrixA.length; idx++) {
            int row = idx;
            executor.submit(new Runnable() {
                public void run() {
                    synchronized (result) {
                        for (int col = 0; col < matrixB[0].length; col++) {
                            for (int idx = 0; idx < matrixB.length; idx++) {
                                result[row][col] = result[row][col] == null ?
                                        matrixA[row][idx].multiply(matrixB[idx][col])
                                        : result[row][col].add(matrixA[row][idx].multiply(matrixB[idx][col]));
                            }
                        }
                        latch.countDown();
                    }
                }
            });
        }
        try {
            latch.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        executor.shutdown();

        return result;
    }

    @Override
    public BigInteger[][] addMatrices(BigInteger[][] matrixA, BigInteger[][] matrixB) {

        for (int row = 0; row < matrixA.length; row++) {
            for (int col = 0; col < matrixA[0].length; col++) {
                matrixA[row][col] = matrixA[row][col].add(matrixB[row][col]);
            }
        }
        return matrixA;
    }

    @Override
    public BigInteger[][] multiplyByScalar(BigInteger[][] matrix, BigInteger scalar) {

        for (int row = 0; row < matrix.length; row++) {
            for (int col = 0; col < matrix[0].length; col++) {
                matrix[row][col] = matrix[row][col].multiply(scalar);
            }
        }
        return matrix;
    }
}
